src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,159 @@
+        } else if (last == null) {
+            last = new Node<>(first, value, null);
+            first.next = last;
+            size++;
+        } else {
+            Node<T> current = last;
+            current.prev.next = current;
+            last = new Node<>(current, value, null);
+            current.next = last;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> current = first;
+            current.next.prev = current;
+            first = new Node<>(null, value, current);
+            current.prev = first;
+            size++;
+        } else {
+            Node<T> nodeAfterCurrent = first;
+            nodeAfterCurrent.next.prev = nodeAfterCurrent;
+            nodeAfterCurrent = findNode(nodeAfterCurrent, index);
+            Node<T> nodeBefore = nodeAfterCurrent.prev;
+            Node<T> current = new Node<>(nodeAfterCurrent.prev, value, nodeAfterCurrent);
+            nodeAfterCurrent.prev = current;
+            nodeBefore.next = current;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = first;
+        currentNode.next.prev = currentNode;
+        currentNode = findNode(currentNode, index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeSet = first;
+        nodeSet.next.prev = nodeSet;
+        nodeSet = findNode(nodeSet, index);
+        T oldValue = nodeSet.value;
+        nodeSet.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T result;
+        if (index == 0) {
+            result = first.value;
+            first = first.next;
+            size--;
+        } else if (index == size - 1) {
+            result = last.value;
+            last.prev.next = null;
+            last = last.prev;
+            size--;
+        } else {
+            Node<T> nodeDeleted = first;
+            nodeDeleted.next.prev = nodeDeleted;
+            nodeDeleted = findNode(nodeDeleted, index);
+            result = nodeDeleted.value;
+            nodeDeleted.prev.next = nodeDeleted.next;
+            nodeDeleted.next.prev = nodeDeleted.prev;
+            size--;
+        }
+        return result;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int count = 0;
+        Node<T> nodeRemoved = first;
+        while (count != size && nodeRemoved.value != object) {
```

Comment:
why do we compare objects through `!=` ?
We need just go through all nodes from the list and compare values

Reply:
I thought that that is exactly good because object is T value  and  node.value is also T value.  And != in case of Null. But not sure.